src/inmanta/server/validate_filter.py

```diff
@@ -34,89 +34,141 @@ def __init__(self, message: str, *args: object) -> None:
+    """
+    if isinstance(v, list):
+        return more_itertools.one(
+            v,
+            too_short=ValueError("Empty filter provided"),
+            too_long=ValueError(f"Multiple values provided for filter: {v}"),
+        )
+
+    return v
 
 
-T = TypeVar("T", bound=FilterModel)
+def parse_range_value_to_date(single_constraint: str, value: str) -> datetime.datetime:
+    try:
+        datetime_obj: datetime.datetime = dateutil.parser.isoparse(value)
+    except ValueError:
+        raise ValueError("Invalid range constraint %s: '%s' is not a valid datetime" % (single_constraint, value))
+    else:
+        return datetime_obj if datetime_obj.tzinfo is not None else datetime_obj.replace(tzinfo=datetime.timezone.utc)
 
 
-class FilterValidator(ABC, Generic[T]):
+def parse_range_operator(v: object) -> Optional[List[Tuple[RangeOperator, datetime.datetime]]]:
     """
-    This class provides methods to validate and process filters as received via the API.
+    Transform list of "<lt|le|gt|ge>:<x>" constraint specifiers to typed objects.
     """
 
-    @property
-    @abstractmethod
-    def model(self) -> Type[T]:
-        raise NotImplementedError()
-
-    def validate_filters(self, filter: Dict[str, List[str]]) -> T:
-        """
-        Validate filters as received via the API and return a corresponding pydantic structure.
-        """
+    def transform_single(single: str) -> Tuple[RangeOperator, datetime.datetime]:
+        split: List[str] = single.split(":", maxsplit=1)
+        if len(split) != 2:
+            raise ValueError("Invalid range constraint %s, expected '<lt|le|gt|ge>:<x>`" % single)
+        operator: RangeOperator
         try:
-            return self.model(**filter)
-        except ValidationError as e:
-            raise InvalidFilter(f"Filter validation failed: {str(e)}, values provided: {str(filter)}") from e
+            operator = RangeOperator.parse(split[0])
+        except ValueError:
+            raise ValueError("Invalid range operator %s in constraint %s, expected one of lt, le, gt, ge" % (split[0], single))
+        bound = parse_range_value_to_date(single, split[1])
+        return (operator, bound)
+
+    if v is None:
+        return None
+
+    if isinstance(v, str):
+        return [transform_single(v)]
+
+    if isinstance(v, list) and all(isinstance(x, str) for x in v):
+        return [transform_single(x) for x in v]
+
+    raise ValueError(f"value is not a valid list of range constraints: {str(v)}")
 
+
+class Filter(ABC):
     @abstractmethod
-    def process_filters(self, filter: Dict[str, List[str]]) -> Dict[str, QueryFilter]:
-        """
-        Processes filters and returns a structured query filter object.
+    def to_query_type(self) -> Optional[Tuple[QueryType, object]]:
+        pass
 
-        :raises InvalidFilter: The supplied filter is invalid.
-        """
-        raise NotImplementedError()
 
+class BooleanEqualityFilter(BaseModel, Filter):
+    field: Optional[bool]
```

Comment:
I would expect this to be the name of a field, so not a boolean?

Reply:
The `Filter` subclasses validate the provided filters with pydantic and describe how to convert them to a `QueryType` which can be used in a database query. (I added this to the docstring)

This means that `field` receives the value to be validated (and the name of the field is set in the `allowed_filters` mapping of the `FilterValidator` classes). This can be simplified when we won't need to support python 3.6 anymore, with generic models